enterprise/views.py

```diff
@@ -1137,3 +1119,95 @@ def post(self, request, enterprise_uuid, program_uuid):
+        return enterprise_customer_uuid, course_run_id or program_uuid
+
+    def eligible_for_direct_audit_enrollment(self, request, enterprise_customer, course_run_id):
+        """
+        Return whether a request is eligible for direct audit enrollment for a particular enterprise customer.
+
+        We check for the following criteria:
+        - The `audit` query parameter.
+        - The user's being routed to the course enrollment landing page.
+        - The customer's catalog contains the course in question.
+        - The audit track is an available mode for the course.
+        """
+        # Return it in one big statement to utilize short-circuiting behavior. Avoid the API call if possible.
+        return request.GET.get('audit') and \
+            request.path == self.COURSE_ENROLLMENT_VIEW_URL.format(enterprise_customer.uuid, course_run_id) and \
+            enterprise_customer.catalog_contains_course(course_run_id) and \
+            EnrollmentApiClient().has_course_mode(course_run_id, 'audit')
+
+    def redirect(self, request, *args, **kwargs):
+        """
+        Redirects to the appropriate view depending on where the user came from.
+        """
+        enterprise_customer_uuid, resource_id = RouterView.get_path_variables(**kwargs)
+        # Replace enterprise UUID and resource ID with '{}', to easily match with a path in RouterView.VIEWS. Example:
+        # /enterprise/fake-uuid/course/course-v1:cool+course+2017/enroll/ -> /enterprise/{}/course/{}/enroll/
+        path = re.sub('{}|{}'.format(enterprise_customer_uuid, re.escape(resource_id)), '{}', request.path)
+        return self.VIEWS[path].as_view()(request, *args, **kwargs)
+
+    @method_decorator(enterprise_login_required)
+    @method_decorator(force_fresh_session)
+    def get(self, request, *args, **kwargs):
+        """
+        Run some custom GET logic for Enterprise workflows before routing the user through existing views.
+
+        In particular, before routing to existing views:
+        - Look to see whether a request is eligible for direct audit enrollment, and if so, directly enroll the user.
+        """
+        enterprise_customer_uuid, resource_id = RouterView.get_path_variables(**kwargs)
+        enterprise_customer = get_enterprise_customer_or_404(enterprise_customer_uuid)
+
+        # Ensure that the link is saved to the database prior to making some call in a downstream view
+        # which may need to know that the user belongs to an enterprise customer.
+        with transaction.atomic():
+            enterprise_customer_user, __ = EnterpriseCustomerUser.objects.get_or_create(
+                enterprise_customer=enterprise_customer,
+                user_id=request.user.id
+            )
+
+        # Directly enroll in audit mode if the request in question has full direct audit enrollment eligibility.
+        if self.eligible_for_direct_audit_enrollment(request, enterprise_customer, resource_id):
```

Comment:
one thing i'm wondering about with regards to firing an event in this case is whether it should live in that enroll method or here in the view. when we were discussing the story i had thought of this event as something that is recording the decision the router view was making, and as such, that it would live in the router view.
i don't think you should change your implementation but curious to hear thoughts on this.

Reply:
Yes that's also what I had read in the description actually, so good point. But it turns out that the actual track event is so generic to enterprise course enrollments (it doesn't contain any special information about having gone through *this* view in particular), that it may just make sense to keep it in the `enroll` function. But I'm open to any other viewpoint!

Reply:
^ Examples of specific cases would be: `'url': request.path`.

Reply:
I like recording the general enterprise enrollment event here. In subsequent work, it would also be nice to track a similar event for all enterprise enrollments.

Perhaps we do need to add another event as well though that tracks the fact that the user came through the direct-to-audit pathway. @UmanShahzad Can you add that?

Reply:
@douglashall Do you mean to keep the event tracking function call inside `enroll` but fire another event here with information about this pathway?

Reply:
Almost, I think the other event relevant to the pathway should be in the view as @brittneyexline suggested.

Reply:
agreed. it makes sense to keep the event in enroll, but other things might call that in the future. conversely, we're also interested in the decisions the router view makes, so two events makes sense.

Reply:
@brittneyexline Ok that makes sense!

Reply:
@douglashall @brittneyexline I've added that specific event now before the redirect in this pathway.

Reply:
(805db539b4cef020dac0b29315a8d2f9b8746beb)

Reply:
Hopefully I didn't misunderstand something here, though.